File: mobile/android/app/src/main/kotlin/com/voquill/mobile/keyboard/KeyboardKeySpec.kt
package com.voquill.mobile.keyboard

data class KeyboardKeySpec(
    val id: String,
    val role: KeyboardKeyRole,
    val label: String,
    val flex: Int = 1,
    val value: String? = null,
)

enum class KeyboardKeyRole {
    CHARACTER, SHIFT, DELETE, SPACE, ENTER, MODE, GLOBE, LANGUAGE, OVERFLOW, START_STOP, UNKNOWN;

    companion object {
        fun from(s: String) = when (s) {
            "character" -> CHARACTER
            "shift" -> SHIFT
            "delete" -> DELETE
            "space" -> SPACE
            "enter" -> ENTER
            "mode" -> MODE
            "globe" -> GLOBE
            "language" -> LANGUAGE
            "overflow" -> OVERFLOW
            "startStop" -> START_STOP
            else -> UNKNOWN
        }
    }
}

File: mobile/ios/keyboard/Types/KeyboardKeySpec.swift
import Foundation

enum KeyboardKeyRole: String {
    case character, shift, delete, space, enter, mode, globe, language, overflow, startStop
    static func from(_ s: String) -> KeyboardKeyRole { KeyboardKeyRole(rawValue: s) ?? .character }
}

struct KeyboardKeySpec {
    let id: String
    let role: KeyboardKeyRole
    let label: String
    let flex: Int
    let value: String?

    static func from(_ dict: [String: Any]) -> KeyboardKeySpec? {
        guard let id = dict["id"] as? String,
              let roleStr = dict["role"] as? String,
              let label = dict["label"] as? String else { return nil }
        return KeyboardKeySpec(
            id: id,
            role: .from(roleStr),
            label: label,
            flex: dict["flex"] as? Int ?? 1,
            value: dict["value"] as? String
        )
    }
}
